File: TGAPEngine/Image.cpp
#include "Image.h"

namespace engine{
	Image::Image(GLubyte* data, unsigned int width, unsigned int height, int unit, GLint format, unsigned int filterMode, bool repeat) :texUnit(unit), format(format), filterMode(filterMode)
{
    glEnable(GL_TEXTURE_2D);
this->imageData = data;
this->imageHeight = height;
this->imageWidth = width;
glGenSamplers(1, &this->sampler);
if (filterMode == TG_FILTER_NEAREST){
	glSamplerParameteri(this->sampler, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
	glSamplerParameteri(this->sampler, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
}

if (filterMode == TG_FILTER_LINEAR){
	glSamplerParameteri(this->sampler, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
	glSamplerParameteri(this->sampler, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
}

if (filterMode == TG_FILTER_MIPMAP_LINEAR){
	glSamplerParameteri(this->sampler, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
	glSamplerParameteri(this->sampler, GL_TEXTURE_MIN_FILTER, GL_LINEAR_MIPMAP_LINEAR);
	
}
if (filterMode == TG_FILTER_NEARMIPMAP_LINEAR){
	glSamplerParameteri(this->sampler, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
	glSamplerParameteri(this->sampler, GL_TEXTURE_MIN_FILTER, GL_NEAREST_MIPMAP_LINEAR);

}
if (repeat){
	glSamplerParameteri(this->sampler, GL_TEXTURE_WRAP_S, GL_REPEAT);
	glSamplerParameteri(this->sampler, GL_TEXTURE_WRAP_T, GL_REPEAT);
}
else{
	glSamplerParameteri(this->sampler,GL_TEXTURE_WRAP_S,GL_CLAMP_TO_EDGE);
	glSamplerParameteri(this->sampler,GL_TEXTURE_WRAP_T,GL_CLAMP_TO_EDGE);
}




glGenBuffers(1,&this->pixelBuffer);
glBindBuffer(GL_PIXEL_UNPACK_BUFFER,this->pixelBuffer);
//glPixelStorei(GL_UNPACK_ALIGNMENT,1);
if(format == GL_LUMINANCE) formatSize = 1;

if (format == GL_RED) formatSize = 1;
if(format == GL_RGB) formatSize = 3;
if(format == GL_RGBA) formatSize = 4;
glBufferData(GL_PIXEL_UNPACK_BUFFER,width * height * sizeof(GLubyte) * formatSize,data,GL_DYNAMIC_DRAW);

//glBufferData(GL_PIXEL_UNPACK_BUFFER,width * height * sizeof(GLubyte),NULL,GL_DYNAMIC_DRAW);
//this->mappedBuffer = glMapBuffer(GL_PIXEL_UNPACK_BUFFER,GL_WRITE_ONLY);
//std::memcpy(this->mappedBuffer,data,1024*64);
//glUnmapBuffer(GL_PIXEL_UNPACK_BUFFER);
//this->mappedBuffer = NULL;

glGenTextures(1,&this->handle);
glBindTexture(GL_TEXTURE_2D,this->handle);
//glTexEnvf( GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, GL_MODULATE );

glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_BASE_LEVEL, 0);
if (filterMode == TG_FILTER_MIPMAP_LINEAR){
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAX_LEVEL, 1000);
	glGenerateMipmap(GL_TEXTURE_2D);
}
else{
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAX_LEVEL, 0);
}

if (data != NULL)
    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, width , height, 0,format, GL_UNSIGNED_BYTE, 0);


if ((filterMode == TG_FILTER_MIPMAP_LINEAR) && (data != NULL)){
	glGenerateMipmap(GL_TEXTURE_2D);
}

glBindTexture(GL_TEXTURE_2D, 0);
glBindBuffer(GL_PIXEL_UNPACK_BUFFER,0);

}

void Image::use()
{
    if(useFlag == true){
   // std::cout<<"using texture"<<std::endl;
        glActiveTexture(GL_TEXTURE0+texUnit);
        glBindTexture(GL_TEXTURE_2D,this->handle);
        glBindSampler(texUnit,this->sampler);
    }
}

GLubyte * Image::getImageData()
{
 return this->imageData;
}

void Image::setImageData(GLubyte* imageDatas)
{
    this->imageData = imageDatas;
    //glActiveTexture(GL_TEXTURE0+texUnit);
    if(DEBUG_MODE == 1)
    std::cout<<"binding"<<std::endl;
    glBindBuffer(GL_PIXEL_UNPACK_BUFFER,this->pixelBuffer);
    glBindTexture(GL_TEXTURE_2D,this->handle);
    //glPixelStorei(GL_UNPACK_ALIGNMENT,1);
    if (this->imageData != NULL){
        //std::cout<<"change is made"<<std::endl;
        //glBufferData(GL_PIXEL_UNPACK_BUFFER,this->imageHeight * this->imageWidth * sizeof(GLubyte),imageDatas,GL_DYNAMIC_DRAW);
        if(DEBUG_MODE == 1)
        std::cout<<"sending data to pbo"<<std::endl;
        if(1 == 2){
            this->mappedBuffer = glMapBuffer(GL_PIXEL_UNPACK_BUFFER,GL_WRITE_ONLY);
            std::memcpy(this->mappedBuffer,imageDatas,this->imageHeight * this->imageWidth * sizeof(GLubyte) * formatSize);
            glUnmapBuffer(GL_PIXEL_UNPACK_BUFFER);
            this->mappedBuffer = NULL;
        }else{
            glBufferSubData(GL_PIXEL_UNPACK_BUFFER,0,this->imageHeight * this->imageWidth * sizeof(GLubyte) * formatSize,imageDatas);
        }
        if(DEBUG_MODE == 1)
        std::cout<<"updating texture from pbo"<<std::endl;
        glTexSubImage2D(GL_TEXTURE_2D, 0, 0,0, this->imageWidth, this->imageHeight,format, GL_UNSIGNED_BYTE, 0);
		if (filterMode == TG_FILTER_MIPMAP_LINEAR){
			glGenerateMipmap(GL_TEXTURE_2D);
		}
        //glTexImage2D(GL_TEXTURE_2D, 0, GL_R8, this->imageWidth , this->imageHeight, 0,GL_RED, GL_UNSIGNED_BYTE, 0);
        if(DEBUG_MODE == 1)
        std::cout<<"updating done, unbinding"<<std::endl;

    }
    glBindTexture(GL_TEXTURE_2D, 0);
    glBindBuffer(GL_PIXEL_UNPACK_BUFFER,0);
}

void Image::setSubImageData(GLubyte *imageData, unsigned int bx, unsigned int by, unsigned int ex, unsigned int ey)
{
	GLuint tmp;
	this->imageData = imageData;
	//glActiveTexture(GL_TEXTURE0+texUnit);
	if (DEBUG_MODE == 1)
		std::cout << "binding" << std::endl;
	glBindBuffer(GL_PIXEL_UNPACK_BUFFER, this->pixelBuffer);
	glBindTexture(GL_TEXTURE_2D, this->handle);
	//glPixelStorei(GL_UNPACK_ALIGNMENT,1);
	if (this->imageData != NULL){
		glGenBuffers(1, &tmp);
		glBindBuffer(GL_PIXEL_UNPACK_BUFFER, tmp);
		glBufferData(GL_PIXEL_UNPACK_BUFFER, (ex-bx) * (ey-by) * sizeof(GLubyte) * formatSize, imageData, GL_DYNAMIC_DRAW);
		if (DEBUG_MODE == 1)
			std::cout << "updating texture from pbo" << std::endl;
		glTexSubImage2D(GL_TEXTURE_2D, 0, bx, by, ex - bx, ey - by, format, GL_UNSIGNED_BYTE, 0);
		if (filterMode == TG_FILTER_MIPMAP_LINEAR){
			glGenerateMipmap(GL_TEXTURE_2D);
		}
		//glTexImage2D(GL_TEXTURE_2D, 0, GL_R8, this->imageWidth , this->imageHeight, 0,GL_RED, GL_UNSIGNED_BYTE, 0);
		if (DEBUG_MODE == 1)
			std::cout << "updating done, unbinding" << std::endl;

	}
	glBindTexture(GL_TEXTURE_2D, 0);
	glBindBuffer(GL_PIXEL_UNPACK_BUFFER, 0);
	//glDeleteBuffers(1, &tmp);
}
}

File: CS/TGAPE_CS/Engine/Renderer/Image.cs
using CSGL;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;

namespace TGAPE_CS.Engine.Renderer
{
    public enum FilterMode : uint
    {
        Nearest = OpenGL.GL_NEAREST,
        Linear = OpenGL.GL_LINEAR,
        MipmapLinear = OpenGL.GL_LINEAR_MIPMAP_LINEAR,
        NearMipmapLinear = OpenGL.GL_NEAREST_MIPMAP_LINEAR
    }
    public class Image
    {
        private FilterMode _filterMode;
        private uint _format;
        private uint _formatSize;
        private byte[] _imageData;
        private uint _imageHeight;
        private uint _imageWidth;
        private uint _textureHandle;
        private uint _samplerHandle;
        private uint _pixelBufferHandle;
        private bool useFlag = true;

        public uint texUnit;

        public bool UseFlag
        {
            private get => useFlag;
            set => useFlag = value;
        }
        public byte[] ImageData
        {
            get => _imageData;
            set => SetImageData(value);
        }


        public Image(byte[] data, uint width, uint height, uint unit, uint format, FilterMode filterMode, bool repeat)
        {
            texUnit = unit;
            _format = format;
            _filterMode = filterMode;

            OpenGL.glEnable(OpenGL.GL_TEXTURE_2D);
            _imageData = data;
            _imageHeight = height;
            _imageWidth = width;
            OpenGL.glGenSamplers(1, ref _samplerHandle);
            var magFilter = filterMode == FilterMode.Nearest ? OpenGL.GL_NEAREST : OpenGL.GL_LINEAR;
            var minFilter = (uint)filterMode;
            OpenGL.glSamplerParameteri(_samplerHandle, OpenGL.GL_TEXTURE_MAG_FILTER, (int)magFilter);
            OpenGL.glSamplerParameteri(_samplerHandle, OpenGL.GL_TEXTURE_MIN_FILTER, (int)minFilter);
            OpenGL.glSamplerParameteri(_samplerHandle, OpenGL.GL_TEXTURE_WRAP_S, repeat ? (int)OpenGL.GL_REPEAT : (int)OpenGL.GL_CLAMP_TO_EDGE);
            OpenGL.glSamplerParameteri(_samplerHandle, OpenGL.GL_TEXTURE_WRAP_T, repeat ? (int)OpenGL.GL_REPEAT : (int)OpenGL.GL_CLAMP_TO_EDGE);

            OpenGL.glGenBuffers(1, ref _pixelBufferHandle);
            OpenGL.glBindBuffer(OpenGL.GL_PIXEL_UNPACK_BUFFER, _pixelBufferHandle);
            switch (format)
            {
                case OpenGL.GL_LUMINANCE:
                    {
                        _formatSize = 1;
                        break;
                    }
                case OpenGL.GL_RED:
                    {
                        _formatSize = 1;
                        break;
                    }
                case OpenGL.GL_RGB:
                    {
                        _formatSize = 3;
                        break;
                    }
                case OpenGL.GL_RGBA:
                    {
                        _formatSize = 4;
                        break;
                    }
                default:
                    {
                        _formatSize = 1;
                        break;
                    }
            }
            unsafe
            {
                fixed (void* ptrData = data)
                    OpenGL.glBufferData(OpenGL.GL_PIXEL_UNPACK_BUFFER, (int)width * (int)height * sizeof(byte) * (int)_formatSize, (IntPtr)ptrData, OpenGL.GL_DYNAMIC_DRAW);
            }
            OpenGL.glGenTextures(1, ref _textureHandle);
            OpenGL.glBindTexture(OpenGL.GL_TEXTURE_2D, _textureHandle);
            OpenGL.glTexParameteri(OpenGL.GL_TEXTURE_2D, OpenGL.GL_TEXTURE_BASE_LEVEL, 0);
            if (filterMode == FilterMode.MipmapLinear)
            {
                OpenGL.glTexParameteri(OpenGL.GL_TEXTURE_2D, OpenGL.GL_TEXTURE_MAX_LEVEL, 1000);
                OpenGL.glGenerateMipmap(OpenGL.GL_TEXTURE_2D);
            }
            else
            {
                OpenGL.glTexParameteri(OpenGL.GL_TEXTURE_2D, OpenGL.GL_TEXTURE_MAX_LEVEL, 0);
            }
            if (data != null)
            {
                OpenGL.glTexImage2D(OpenGL.GL_TEXTURE_2D, 0, (int)OpenGL.GL_RGBA8, (int)width, (int)height, 0, format, OpenGL.GL_UNSIGNED_BYTE, IntPtr.Zero);
            }
            if (filterMode == FilterMode.MipmapLinear && data != null)
            {
                OpenGL.glGenerateMipmap(OpenGL.GL_TEXTURE_2D);
            }
            OpenGL.glBindTexture(OpenGL.GL_TEXTURE_2D, 0);
            OpenGL.glBindBuffer(OpenGL.GL_PIXEL_UNPACK_BUFFER, 0);
        }
        public void Use()
        {
            if (UseFlag)
            {
                OpenGL.glActiveTexture(OpenGL.GL_TEXTURE0 + texUnit);
                OpenGL.glBindTexture(OpenGL.GL_TEXTURE_2D, _textureHandle);
                OpenGL.glBindSampler(texUnit, _samplerHandle);
            }
        }
        public void Reject()
        {
            OpenGL.glBindSampler(texUnit, 0);
            OpenGL.glBindTexture(OpenGL.GL_TEXTURE_2D, 0);
            OpenGL.glActiveTexture(OpenGL.GL_TEXTURE0);
        }
        public void SetImageData(byte[] imageData)
        {
            _imageData = imageData;
            OpenGL.glBindBuffer(OpenGL.GL_PIXEL_UNPACK_BUFFER, _pixelBufferHandle);
            OpenGL.glBindTexture(OpenGL.GL_TEXTURE_2D, _textureHandle);
            if (_imageData != null)
            {
                unsafe
                {
                    fixed (void* ptrData = imageData)
                        OpenGL.glBufferSubData(OpenGL.GL_PIXEL_UNPACK_BUFFER, 0,(int)_imageHeight * (int)_imageWidth * sizeof(byte) * (int)_formatSize, (IntPtr)ptrData);
                }
                OpenGL.glTexSubImage2D(OpenGL.GL_TEXTURE_2D, 0, 0, 0, (int)_imageWidth, (int)_imageHeight, _format, OpenGL.GL_UNSIGNED_BYTE, IntPtr.Zero);
                if(_filterMode == FilterMode.MipmapLinear)
                {
                    OpenGL.glGenerateMipmap(OpenGL.GL_TEXTURE_2D);
                }
            }
            OpenGL.glBindTexture(OpenGL.GL_TEXTURE_2D, 0);
            OpenGL.glBindBuffer(OpenGL.GL_PIXEL_UNPACK_BUFFER, 0);
        }
        //TODO:magic numbers I not sure now whar coors are they
        public void SetSubImageData(byte[] imageData, uint bx, uint by, uint ex, uint ey)
        {
            uint tempBuffer = 0;
            _imageData = imageData;
            OpenGL.glBindBuffer(OpenGL.GL_PIXEL_UNPACK_BUFFER, _pixelBufferHandle);
            OpenGL.glBindTexture(OpenGL.GL_TEXTURE_2D, _textureHandle);
            if (_imageData != null)
            {
                OpenGL.glGenBuffers(1, ref tempBuffer);
                OpenGL.glBindBuffer(OpenGL.GL_PIXEL_UNPACK_BUFFER, tempBuffer);

                unsafe
                {
                    fixed (void* ptrData = imageData)
                        OpenGL.glBufferData(OpenGL.GL_PIXEL_UNPACK_BUFFER, (int)((ex - bx) * (ey - by) * sizeof(byte) * _formatSize), (IntPtr)ptrData, OpenGL.GL_DYNAMIC_DRAW);
                }
                OpenGL.glTexSubImage2D(OpenGL.GL_TEXTURE_2D, 0, (int)bx, (int)by, (int)(ex - bx), (int)(ey - by), _format, OpenGL.GL_UNSIGNED_BYTE, IntPtr.Zero);
                if (_filterMode == FilterMode.MipmapLinear)
                {
                    OpenGL.glGenerateMipmap(OpenGL.GL_TEXTURE_2D);
                }
            }
            OpenGL.glBindTexture(OpenGL.GL_TEXTURE_2D, 0);
            OpenGL.glBindBuffer(OpenGL.GL_PIXEL_UNPACK_BUFFER, 0);
        }

    }
}
